lib/perfdata/influxdbchecktask.cpp

```diff
@@ -0,0 +1,149 @@
+}
+
+void InfluxdbCheckTask::ScriptFunc(
+	const Checkable::Ptr& checkable, const CheckResult::Ptr& cr,
+	const Dictionary::Ptr& resolvedMacros, bool useResolvedMacros
+)
+{
+	ServiceState state = ServiceOK;
+	CheckCommand::Ptr commandObj = CheckCommand::ExecuteOverride ? CheckCommand::ExecuteOverride : checkable->GetCheckCommand();
+
+	Host::Ptr host;
+	Service::Ptr service;
+	tie(host, service) = GetHostService(checkable);
+
+	MacroProcessor::ResolverList resolvers;
+
+	if (MacroResolver::OverrideMacros) {
+		resolvers.emplace_back("override", MacroResolver::OverrideMacros);
+	}
+
+	if (service)
+		resolvers.emplace_back("service", service);
+	resolvers.emplace_back("host", host);
+	resolvers.emplace_back("command", commandObj);
+	resolvers.emplace_back("icinga", IcingaApplication::GetInstance());
+
+	String influxdbWriterName = MacroProcessor::ResolveMacros("$influxdbwriter_name$", resolvers, checkable->GetLastCheckResult(),
+		nullptr, MacroProcessor::EscapeCallback(), resolvedMacros, useResolvedMacros);
+
+	String missingQueriesWarning;
+	String missingQueriesCritical;
+	String missingPendingDataBuffersWarn;
+	String missingPendingDataBuffersCrit;
+
+	double queriesWarning = MacroProcessor::ResolveMacros("$influxdbwriter_queries_warning$", resolvers, checkable->GetLastCheckResult(),
+		&missingQueriesWarning, MacroProcessor::EscapeCallback(), resolvedMacros, useResolvedMacros);
+
+	double queriesCritical = MacroProcessor::ResolveMacros("$influxdbwriter_queries_critical$", resolvers, checkable->GetLastCheckResult(),
+		&missingQueriesCritical, MacroProcessor::EscapeCallback(), resolvedMacros, useResolvedMacros);
+
+	double pendingDataBuffersWarn = MacroProcessor::ResolveMacros("$influxdbwriter_pending_queries_warning$", resolvers, checkable->GetLastCheckResult(),
+		&missingPendingDataBuffersWarn, MacroProcessor::EscapeCallback(), resolvedMacros, useResolvedMacros);
+
+	double pendingDataBuffersCrit = MacroProcessor::ResolveMacros("$influxdbwriter_pending_queries_critical$", resolvers, checkable->GetLastCheckResult(),
+		&missingPendingDataBuffersCrit, MacroProcessor::EscapeCallback(), resolvedMacros, useResolvedMacros);
+
+	if (resolvedMacros && !useResolvedMacros)
+		return;
+
+	if (influxdbWriterName.IsEmpty()) {
+		ReportInfluxdbCheck(checkable, commandObj, cr, "Attribute 'influxdbwriter_name' must be set.");
+		return;
+	}
+
+	InfluxdbWriter::Ptr influxdbWriter = InfluxdbWriter::GetByName(influxdbWriterName);
+
+	if (!influxdbWriter) {
+		ReportInfluxdbCheck(checkable, commandObj, cr, "InfluxdbWriter '" + influxdbWriterName + "' object does not exist.");
+		return;
+	}
+
+	if (influxdbWriter->IsPaused()) {
+		ReportInfluxdbCheck(checkable, commandObj, cr, "InfluxdbWriter is temporarily disabled on this cluster instance.", state);
+		return;
+	}
+
+	auto errorMessage = influxdbWriter->GetLastErrorMessage();
+	if (!errorMessage.IsEmpty()) {
+		ReportInfluxdbCheck(checkable, commandObj, cr, errorMessage, ServiceCritical);
+		return;
+	}
```

Comment:
How should you add here Perfdata when the writer either can't connect with InfluxDB or an exception is raised? The only useful information regarding why this is happening is cached and given out as output.

Reply:
I was thinking that if `SetLastErrorMessage` is called in `InfluxdbWriter::ExceptionHandler`, this should not necessarily imply that everything is broken. At least the `WorkQueue` seems to go on after an exception with other jobs.

Reply:
Yes, I do agree with you but without connecting to InfluxDB successfully there is no useful data to be reported and when the writer has connected successfully the ``SetLasstErrorMessage()`` is called again and set to empty. So, even if it continues with other WorkQueue tasks after raising an exception, it will anyway come to the ``Flush()`` method where the LastErrorMessage is set again.

Reply:
I wouldn't say that the data is not useful. If there's an error, it would be perfectly fine to see the rate drop in the perfdata instead of having a gap there. Also this doesn't require any major changes to the code, just move `cr->SetPerformanceData()` and its dependencies above this block.